pandas/core/arrays/integer.py

```diff
@@ -377,14 +377,28 @@ def __getitem__(self, item):
             return self._data[item]
         return type(self)(self._data[item], self._mask[item])
 
-    def _coerce_to_ndarray(self):
+    def _coerce_to_ndarray(self, dtype=None, na_value=libmissing.NA):
         """
         coerce to an ndarary of object dtype
         """
-
         # TODO(jreback) make this better
-        data = self._data.astype(object)
-        data[self._mask] = self._na_value
+        if dtype is None:
+            dtype = object
+        elif is_float_dtype(dtype) and na_value is libmissing.NA:
+            # XXX: Do we want to implicitly treat NA as NaN here?
+            # We should be deliberate in this decision.
+            na_value = np.nan
+
+        data = self._data.astype(dtype)
+
+        if (
+            is_integer_dtype(dtype)
+            and na_value is libmissing.NA
+            and not self._mask.any()
+        ):
+            return data
```

Comment:
I believe this retains the behavior on master of

```
np.asarray(IntegerArray, dtype=int)
```

succeeding if there are no NA values, and failing otherwise. Is that the behavior we want? If so, I'll ensure we have a better error message.

Reply:
> succeeding if there are no NA values, and failing otherwise. Is that the behavior we want?

I think so, yes.